vms/platformvm/validators/manager.go

```diff
@@ -170,56 +171,105 @@ func (m *manager) GetValidatorSet(ctx context.Context, height uint64, subnetID i
 		return nil, ErrMissingValidator
 	}
 
-	currentSubnetValidatorList := currentSubnetValidators.List()
-	vdrSet := make(map[ids.NodeID]*validators.GetValidatorOutput, len(currentSubnetValidatorList))
-	for _, vdr := range currentSubnetValidatorList {
-		primaryVdr, ok := currentPrimaryNetworkValidators.Get(vdr.NodeID)
+	currentSubnetValidatorList := currentSubnetValidatorSet.List()
+	currentSubnetValidators := make([]*maskedValidator, 0, len(currentSubnetValidatorList))
+	currentSubnetValidatorIndices := make(map[ids.NodeID]int)
+	for i, validator := range currentSubnetValidatorList {
+		primaryVdr, ok := currentPrimaryNetworkValidators.Get(validator.NodeID)
 		if !ok {
 			// This should never happen
-			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, vdr.NodeID)
-		}
-		vdrSet[vdr.NodeID] = &validators.GetValidatorOutput{
-			NodeID:    vdr.NodeID,
-			PublicKey: primaryVdr.PublicKey,
-			Weight:    vdr.Weight,
+			return nil, fmt.Errorf("%w: %s", ErrMissingValidator, validator.NodeID)
 		}
+
+		currentSubnetValidators = append(currentSubnetValidators, &maskedValidator{
+			GetValidatorOutput: &validators.GetValidatorOutput{
+				NodeID:    validator.NodeID,
+				PublicKey: primaryVdr.PublicKey,
+				Weight:    validator.Weight,
+			},
+			masked: false,
+		})
+		currentSubnetValidatorIndices[validator.NodeID] = i
 	}
 
+	validatorsFromDiffs := make([]*maskedValidator, 0)
+	validatorsFromDiffsIndices := make(map[ids.NodeID]int)
 	for diffHeight := lastAcceptedHeight; diffHeight > height; diffHeight-- {
-		err := m.applyValidatorDiffs(vdrSet, subnetID, diffHeight)
-		if err != nil {
+		if validatorsFromDiffs, err = m.applyValidatorDiffs(
+			currentSubnetValidators,
+			currentSubnetValidatorIndices,
+			validatorsFromDiffs,
+			validatorsFromDiffsIndices,
+			subnetID,
+			diffHeight,
+		); err != nil {
 			return nil, err
 		}
 	}
 
+	// The current validator set is guaranteed to be sorted. Sort the set of
+	// validators calculated from applying diffs, so we can merge the two sorted
+	// lists.
+	sort.Slice(validatorsFromDiffs, func(i, j int) bool {
+		return validatorsFromDiffs[i].less(validatorsFromDiffs[j])
+	})
+
+	result := merge(currentSubnetValidators, validatorsFromDiffs)
```

Comment:
IIRC, insertion sort is the best algo for nearly sorted arrays, which I think it's our case, if we consider an initially sorted validators list and not-too-deep height. Also prolly the whole implementation will be simpler, ssince IIUC, validator's weight change in diffs won't alter the order. Only insertion and removal will change the order.
It what I understand above is true, I'd go for a simpler insertion sort implementation